Implement paintings

Postremus opened the pull request:
This PR attempts to implement paintings.

GlowHangingEntity and HangingStore are reused for this, since the facing value is the same as for itemframes.

Stores the "Motive" of a painting in CamelCase, like vanilla does it.

When "Painting.setFacing" or "Painting.setArt" is called on paper, a DestroyEntities and SpawnPainting Message is sent. A new id is used for spawning, which probably means a new entity is created.
Changing of the entity id can not be easily done here, and does not make much sense, therefore not implemented.

getHeight() and getWidth() return regardless of the art size 0.5 each, which is the same as paper.

The BoundingBox x and z of a painting is defined as "type width or 0.0625 (depth) ". I had to adjust the EntityBoundingBox to make it possible to set width and depth separately.

A painting has multiple important locations, each relative to the topleftcorner of the painting:
- center of art: Defined as (max(0, artWidth / 2 - 1), artHeight / 2)). This one is needed for spawning this entity. It is also used as center of rotation.
- center of painting entity: Defined as (artWidth / 2, artHeight / 2). This one is the center, and location, of the entity.

related to #504

Postremus committed:
Cleanup Itemframes
Move facing attribute into new super class
create new enum hangingface, to replace facingNumber
call some more events

Postremus committed:
Add paintings
load and save paintings

Postremus committed:
Spawn paintings when using item

Postremus committed:
Add recipe for paintings

Postremus committed:
Implement set/get of facing and art
rotated art is created by despawning the old painting, and spawning another one with the new facing
fixed recipes
use correct art name

Postremus committed:
Try not to spawn paintings in blocks
some initial cleanup

Postremus committed:
Choose more appropriate painting size
some more cleanup

Postremus committed:
respect entities bounding boxes better

Postremus committed:
respawn painting after rotation
respect force flag

Postremus committed:
some cleanup

Postremus committed:
make naming of motive in nbt consistent with vanilla

Postremus committed:
use random art when placed by hand

Postremus committed:
destroy painting on attack

Postremus committed:
correct location of entity
location is not the same as the center
the center is defined as (max(0, width / 2 - 1, height / 2)
the location is defined as (width / 2, height / 2)
also set the yaw of the paintings location

Postremus committed:
fix bounding box size

also some cleanup

Postremus committed:
Move the center of the image along

play place sound

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowPainting.java:
```diff
@@ -0,0 +1,334 @@
+        super.pulse();
+
+        if (ticksLived % 11 == 0) {
+
```
Whitespace

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowPainting.java:
```diff
@@ -0,0 +1,334 @@
+
+    @Override
+    public double getWidth() {
+        return 0.5;
```
Add a note about this value being static (like you mentioned in the PR description)

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowPainting.java:
```diff
@@ -0,0 +1,334 @@
+    public void pulse() {
+        super.pulse();
+
+        if (ticksLived % 11 == 0) {
```
Where does the `11` value come from? Do paintings only update every 0.55 seconds?

Postremus replied:
I took the value from the itemframes, since they are hanging entities as well. I will recheck this.

Postremus replied:
Did some testing on vanilla using a clock.
The hanging entity is only destroyed once `ticksLived % 5 seconds == 0`.
This is true for paintings as well as item frames.  

On paper, this holds true as well.
https://gist.github.com/Postremus/5358c837099817bbfd98bcdbad01ee1e

Postremus replied:
I adjusted these checks to be more in line with vanilla.

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowPainting.java:
```diff
@@ -0,0 +1,334 @@
+        return true;
+    }
+
+    private void respawn() {
```
Add documentation about this method (sends destroy and then spawn messages).
I would rename this method `refresh` and make it available publicly for plugins wanting to use it.

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowPainting.java:
```diff
@@ -0,0 +1,334 @@
+            .forEach(p -> p.getSession().sendAll(messages));
+    }
+
+    public void setArtInternal(Art art) {
```
Add documentation about this method because it is public.

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowPainting.java:
```diff
@@ -0,0 +1,334 @@
+        // not affected by physics
+    }
+
+    public boolean isObstructed() {
```
Add documentation about this method because it is public.

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowPainting.java:
```diff
@@ -0,0 +1,334 @@
+            current = current.getBlock().getRelative(BlockFace.DOWN).getLocation();
+
+            // reset x and z
+            current.subtract(right.getModX() * art.getBlockWidth(), 0, right.getModZ() * art.getBlockWidth());
```
Since you are using `subtract`, make sure the locations are cloned before updating them, IIRC it can update the block's location too (unless that only happens on Spigot?)

Postremus replied:
`getTopLeftCorner()`, which is used to obtain the starting location, returns a cloned location.  
`GlowBlock.getLocation()` always returns a new Location, therefore the location of the block is not updated.

aramperes commented on src/main/java/net/glowstone/net/message/play/entity/SpawnPaintingMessage.java:
```diff
@@ -1,12 +1,14 @@
 public final class SpawnPaintingMessage implements Message {
 
     private final int id;
+    private final UUID uuid;
```
Rename this field to `uniqueId`.

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowPainting.java:
```diff
@@ -0,0 +1,334 @@
+    }
+
+    @Override
+    public boolean setArt(Art art) {
```
Check if `art` is not null, default to Kebab if it is.

Postremus replied:
Paper just throws a NPE in this case: https://gist.github.com/Postremus/a2efa23ead7915e95183f6d93610454c  
I changed setArt and setFacing to do the same.

aramperes replied:
I wouldn't throw the exception. That's a bug with their implementation and is not an API design choice - use notnull validation here.

aramperes commented:
This will likely be our last feature for the `2017.7` release which is set to Monday/Tuesday. I'd like to thank you for all of your contributions you've done recently!

Postremus committed:
Update documentation
remove duplicate calls to updateBoundingBox

Postremus committed:
The center of the art must be stored as Pos
instead of the location
this mimics the vanilla behaviour, and also fixes some dislocations upon loading of the world

also fix bounding boxes

Postremus committed:
Only check every 5 seconds if the hanging entity can survive

Momothereal committed:
Documentation adjustments

aramperes commented on src/main/java/net/glowstone/entity/objects/GlowPainting.java:
```diff
@@ -0,0 +1,365 @@
+            current.subtract(right.getModX() * art.getBlockWidth(), 0, right.getModZ() * art.getBlockWidth());
+        }
+
+        return !this.world.getEntityManager().getEntitiesInside(this.boundingBox, this).isEmpty();
```
What happens if a player 'collides' with the painting? Wouldn't it be destroyed here?

Postremus replied:
Changed this so only hanging entities are checked for.

Postremus committed:
son, never forget, hanging entities are your kryptonite

Postremus committed:
Merge remote-tracking branch 'mypp/#504-painting' into #504-painting
